File: web/src/lastOpenedSession.ts
const LAST_OPENED_SESSION_KEY = "piss:last-opened-session";
const SESSION_OPEN_HISTORY_KEY = "piss:session-open-history";
const MAX_SESSION_ID_LENGTH = 512;
const MAX_SESSION_HISTORY_LENGTH = 200;

type SessionStorage = Pick<Storage, "getItem" | "setItem" | "removeItem">;
export type SessionOpenHistory = Readonly<Record<string, number>>;

export function readSessionOpenHistory(storage: SessionStorage = localStorage): SessionOpenHistory {
  try {
    const value: unknown = JSON.parse(storage.getItem(SESSION_OPEN_HISTORY_KEY) ?? "{}");
    if (!value || typeof value !== "object" || Array.isArray(value)) return {};
    return Object.fromEntries(Object.entries(value)
      .filter(([sessionId, openedAt]) => sessionId.length > 0
        && sessionId.length <= MAX_SESSION_ID_LENGTH
        && typeof openedAt === "number"
        && Number.isFinite(openedAt)
        && openedAt >= 0)
      .sort((left, right) => (right[1] as number) - (left[1] as number))
      .slice(0, MAX_SESSION_HISTORY_LENGTH));
  } catch {
    return {};
  }
}

export function readLastOpenedSession(storage: SessionStorage = localStorage): string | undefined {
  try {
    const sessionId = storage.getItem(LAST_OPENED_SESSION_KEY);
    if (!sessionId || sessionId.length > MAX_SESSION_ID_LENGTH) {
      if (sessionId) storage.removeItem(LAST_OPENED_SESSION_KEY);
      return;
    }
    return sessionId;
  } catch {
    return;
  }
}

export function writeLastOpenedSession(
  sessionId: string | undefined,
  storage: SessionStorage = localStorage,
  openedAt = Date.now(),
): SessionOpenHistory {
  const history = readSessionOpenHistory(storage);
  try {
    if (!sessionId) {
      storage.removeItem(LAST_OPENED_SESSION_KEY);
      return history;
    }
    storage.setItem(LAST_OPENED_SESSION_KEY, sessionId);
    const nextHistory = Object.fromEntries([
      [sessionId, openedAt],
      ...Object.entries(history).filter(([id]) => id !== sessionId),
    ].slice(0, MAX_SESSION_HISTORY_LENGTH));
    storage.setItem(SESSION_OPEN_HISTORY_KEY, JSON.stringify(nextHistory));
    return nextHistory;
  } catch {
    // Storage may be unavailable or full in private browsing modes.
    return history;
  }
}

File: web/last_opened_session.ml
open! Core
open Js_of_ocaml

let storage_key = "piss:last-opened-session"
let max_session_id_length = 512
let storage () = Js.Unsafe.get (Js.Unsafe.inject Dom_html.window) "localStorage"

let remove () =
  ignore
    (Js.Unsafe.meth_call (storage ()) "removeItem"
       [| Js.Unsafe.inject (Js.string storage_key) |])

let read () =
  try
    let stored =
      Js.Unsafe.meth_call (storage ()) "getItem"
        [| Js.Unsafe.inject (Js.string storage_key) |]
    in
    let present : bool Js.t =
      Js.Unsafe.fun_call
        (Js.Unsafe.get (Js.Unsafe.inject Dom_html.window) "Boolean")
        [| stored |]
      |> Js.Unsafe.coerce
    in
    let value =
      if Js.to_bool present then Some (Js.to_string (Js.Unsafe.coerce stored))
      else None
    in
    match value with
    | Some session_id
      when (not (String.is_empty session_id))
           && String.length session_id <= max_session_id_length ->
        Some session_id
    | Some _ ->
        remove ();
        None
    | None -> None
  with _ -> None

let write = function
  | None -> ( try remove () with _ -> ())
  | Some session_id -> (
      if
        (not (String.is_empty session_id))
        && String.length session_id <= max_session_id_length
      then
        try
          ignore
            (Js.Unsafe.meth_call (storage ()) "setItem"
               [|
                 Js.Unsafe.inject (Js.string storage_key);
                 Js.Unsafe.inject (Js.string session_id);
               |])
        with _ -> ())
